[REVIEW] Enable multiple column groups in TargetEncoding op

rjzamora opened the pull request:
Closes #261 

This enables the TargetEncoding op to accomodate multiple column groups at once.  All coolumn groups must use the same continuous target column (we can introduce a follow-up PR to loosen this requirement if/when necessary).

Since there can now be multiple categorical-column groups, multi-column groups must be specified within two layers of square brackets.  As mentioned in the docstring:

```
    cat_groups : list of column-groups
        Columns, or column groups, to target encode. A single encoding
        will include multiple categorical columns if the column names are
        enclosed within two layers of square brackets.  For example,
        `["a", "b"]` means "a" and "b" will be separately encoded, while
        `[["a", "b"]]` means they will be encoded as a single group.
        Note that the same column can be used for multiple encodings.
        For example, `["a", ["a", "b"]]` is valid.
```

<!--

Thank you for contributing to NVTabular :)

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

4. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

5. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present) and replace
   it with `[REVIEW]`. If assistance is required to complete the functionality,
   for example when the C/C++ code of a feature is complete but Python bindings
   are still required, then add the label `[HELP-REQ]` so that others can triage
   and assist. The additional changes then can be implemented on top of the
   same PR. If the assistance is done by members of the rapidsAI team, then no
   additional actions are required by the creator of the original PR for this,
   otherwise the original author of the PR needs to give permission to the
   person(s) assisting to commit to their personal fork of the project. If that
   doesn't happen then a new PR based on the code of the original PR can be
   opened by the person assisting, which then will be the PR that will be
   merged.

6. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please do not
   rebase your branch on master/force push/rewrite history, doing any of these
   causes the context of any comments made by reviewers to be lost. If
   conflicts occur against master they should be resolved by merging master
   into the branch used for making the pull request.

Many thanks in advance for your cooperation!

-->

jperez999 commented:
rerun tests

jperez999 commented:
rerun tests

rjzamora commented:
@bschifferer - Let me know if these changes address your needs.  Sorry this was missing from the original PR!

rjzamora commented:
*Update*:  After running through some performance experiments, I am noticing that we may need to avoid leaving the `"__fold__"` column in `ddf` after returning from `GroupbyStatistics`...  In #248 , we changed the Workflow to update the underlying `ddf` after each statistics operation.  This can lead to more memory pressure than I originally anticipated.  The scheduler/workers will no longer "forget" about the ddf transformations performed during the statistics-gathering step, but now try to persist the changes in device memory.  For large datasets, this is only possible with device-to-host spilling (which can be slow)...  We may want to make the fold indices deterministic, and just add the `"__fold__"` column again during each TargetEncoding transform task. 
cc @benfred

benfred commented:
> _Update_: After running through some performance experiments, I am noticing that we may need to avoid leaving the `"__fold__"` column in `ddf` after returning from `GroupbyStatistics`... In #248 , we changed the Workflow to update the underlying `ddf` after each statistics operation. This can lead to more memory pressure than I originally anticipated. The scheduler/workers will no longer "forget" about the ddf transformations performed during the statistics-gathering step, but now try to persist the changes in device memory. For large datasets, this is only possible with device-to-host spilling (which can be slow)... We may want to make the fold indices deterministic, and just add the `"__fold__"` column again during each TargetEncoding transform task.
> cc @benfred

Thanks for the heads up!  I could see this causing issues with larger datasets. Are there any other transforms like this happening in other ops?

rjzamora commented:
> Thanks for the heads up! I could see this causing issues with larger datasets. Are there any other transforms like this happening in other ops?

TargetEncoding is the only op that requires its statistics-op dependency to modify the ddf.  Which is why I would like to roll back this behaviour if possible...

I am currently experimenting with a slightly different approach: Passing the partition index to every transform op, so that a deterministic fold mapping can be reconstructed at both statistics-gathering and transform time.  Unfortunately, this means we will need to plumb a `partition_index` argument for all the `apply_op` and `op_logic` definitions.  We can avoid this if we are okay with the same fold asignment being the same in every partition (which actually seems to be the case now anyway), but that seems like a rough limitation to me.  Also, since the `"__fold__"` column will no longer persist after GroupbyStatistics, we will need an additional argument for the user to specify if cross-validation should be used at transform time (e.g. `training_data=True`)

rjzamora commented:
**Update 2**:  I didn't find it to be much of an advantage to avoid updating ddf in `GroupbyStatistics`, and adding the partition index to all transofrmation options was getting a little "messy" (in my opinion).   For reference, I did try it out [in this commit](https://github.com/rjzamora/NVTabular/commit/a85c8ed14a7163a584452bfc091185fd46dc2dcd).  Given what I learned, I made a few different optimizations:

- Added `target_mean` argument to allow users to specify the global mean of the target continuous column (if it is already known).  It turns out that the dependency on the mean is the primary cause of the memory-pressure issues I was seeing before (the addition of a `"__fold__"` column is less important than I had thought).
- Modified `_write_gb_stats` to avoid any concatenation, and to avoid host-to-device transfer when the `on_host=True`.  More-specifically, we now loop through the incomming DataFrame's and append them to the same file with cudf or pyarrow, depending on the format of the DFs.  We no longer concatenate and convert to cudf, because we do not need to sort for the groupby statistics (just the uniques used in Categorify).  If, in the future, we enable the same `GroupbyStatistics` instance to cover both `Categorify` and other transformations, we will need to ensure that the data is sorted during/before the `Categorify` encoding step (if/when the `searchsorted` method is used).

rjzamora commented:
@benfred - I'm not sure I will have time to add any more optimizations here before 0.2 release.  I'll plan to merge this once the PR has your blessing.

benfred commented on nvtabular/ops/categorify.py:
```diff
@@ -457,24 +460,43 @@ def _get_aggregation_type(col):
+            n_writes += 1
+
+    # Close writer and return path
+    pwriter.close()
```
Maybe we should move the ```pwriter.close()``` to after the ```if n_writes == 0``` check?  When there is no data written out, and on_host = True - I believe this line will throw a ```AttributeError: 'NoneType' object has no attribute 'close'``` exception

benfred reviewed (approved):
LGTM -
